### PR TITLE
Extract Credential Record abstraction

### PR DESCRIPTION
**Pull request by emlun**
This is a step towards resolving #1510 and #1556, but is substantive enough to warrant review on its own. This depends on PR #1771 and is a prerequisite of PR #1774.

I open this as a "draft" PR because it has PR #1771 as its base branch, to isolate the diff to the changes unique to this PR. I will re-target this PR to the main branch before merging.

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/webauthn/pull/1773.html" title="Last updated on Sep 22, 2022, 2:18 PM UTC (f6479ef)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/webauthn/1773/573b1c2...f6479ef.html" title="Last updated on Sep 22, 2022, 2:18 PM UTC (f6479ef)">Diff</a>

### CODE REVIEW

File: index.bs
Comment (timcappalli) [inline index.bs:1160]: Should we recommend that if the device public key is requested and returned, it should be stored with or with a reference to the credential record?

Comment (emlun) [reply index.bs:1160]: Hm, good point. I'm not sure it's appropriate to call out specific extensions here, but we should at least point out that extensions might add items that need to be stored. Perhaps as an extension point similar to how extensions add fields to `partial dictionary AuthenticationExtensionsClientOutputs` etc?

File: index.bs
Comment (Firstyear) [inline index.bs:1150]: Reality seems to be the opposite, seems to be that chrome doesn't properly assign all the transports for a credential during registration with getTransports, so removing (clearing) this value is actually safe 😅

File: index.bs
Comment (Firstyear) [reply index.bs:1160]: we probably also need to store the attestation ca chain or related attestation metadata to assist with DPK here. We already store this in webauthn-rs, here: https://github.com/kanidm/webauthn-rs/blob/master/webauthn-rs-core/src/interface.rs#L245

File: index.bs
Comment (agl) [reply index.bs:1150]: Chrome had a bug where it was overconfident that it knew a complete set of transports. Chrome 105 will report an empty set unless it has transports from the getInfo or attestation certificate.

File: index.bs
Comment (agl) [reply index.bs:1160]: (Irrespective of DPK, some enterprise RPs certainly like to store the attestation for record keeping.)

File: index.bs
Comment (agl) [inline index.bs:5141]: (The definition, above, also includes a `type` member.)

**Review by agl — APPROVED** [review]
LGTM with nits.

File: index.bs
Comment (emlun) [reply index.bs:1150]: Setting `transports` to empty in `PublicKeyCredentialDescriptor` is equivalent to omitting it, so yes, that should always work in the sense of not preventing use of a credential. This could make the user experience slightly worse if the user only has `internal`-only credentials, though, as they might get prompted to connect an external authenticator they don't have.

However if you remove items from `transports` but it remains non-empty, then that could prevent using a credential if the client assumes the credential requires an unavailable transport (say, if `getTransports()` returns `["usb", "nfc"]` and you remove `"usb"`, then pass `transports: ["nfc"]` on a device with no NFC antenna).

**Review by akshayku — APPROVED** [review]

**Review by selfissued — APPROVED** [review]
LGTM

**Review by timcappalli — APPROVED** [review]
lgtm

File: index.bs
Comment (emlun) [reply index.bs]: Should be fixed by a5710f6, please re-review!

**Comment by emlun** [comment]
On the 2022-09-21 WG call we said to merge this after I'd fixed @agl's [review comment](https://github.com/w3c/webauthn/pull/1773#discussion_r950739605), but I'd forgotten there was another [unresolved comment thread](https://github.com/w3c/webauthn/pull/1773#discussion_r918219321). That led to a slightly more substantial change, so I'll give @timcappalli an opportunity to re-review before we merge this.

File: index.bs
Comment (timcappalli) [inline index.bs:1173]: @emlun now that attestation is possible on getAssertion, do we need additional language here to address that?

Comment (emlun) [reply index.bs:1173]: Yes, but I'd like to tie that into #1663 instead since that's where that change will be introduced. I intend to open a meta-PR to update #1663 to use the new credential record abstraction once this is merged (unless #1663 gets merged before I can get to it :smile:).

**Comment by timcappalli** [comment]
changes lgtm

**Comment by emlun** [comment]
Thanks @timcappalli (and everyone else)!

**Pull request merged**
